packages/create-amplify/src/package-manager-controller/package_manager_controller_factory.ts

```diff
@@ -0,0 +1,157 @@
+import { PnpmPackageManagerController } from './pnpm_package_manager_controller.js';
+import { YarnClassicPackageManagerController } from './yarn_classic_package_manager_controller.js';
+import { YarnModernPackageManagerController } from './yarn_modern_package_manager_controller.js';
+import {
+  NpmProjectInitializer,
+  PnpmProjectInitializer,
+  YarnClassicProjectInitializer,
+  YarnModernProjectInitializer,
+} from '../project-initializer/index.js';
+
+export type DependencyType = 'dev' | 'prod';
+
+/**
+ *
+ */
+export abstract class PackageManagerController {
+  abstract installDependencies: (
+    packageNames: string[],
+    type: DependencyType
+  ) => Promise<void>;
+}
+
+/**
+ * packageManagerControllerFactory
+ */
+export class PackageManagerControllerFactory {
+  private readonly existsSync = _existsSync;
+  packageManagerName: PackageManagerName;
+  private readonly packageManagers: PackageManagers = {
+    npm: {
+      name: 'npm',
+      executable: 'npm',
+      binaryRunner: 'npx',
+      installCommand: 'install',
+      lockFile: 'package-lock.json',
+      initDefault: ['init', '--yes'],
+    },
+    'yarn-classic': {
+      name: 'yarn-classic',
+      executable: 'yarn',
+      binaryRunner: 'yarn',
+      installCommand: 'add',
+      lockFile: 'yarn.lock',
+      initDefault: ['init', '--yes'],
+    },
+    'yarn-modern': {
+      name: 'yarn-modern',
+      executable: 'yarn',
+      binaryRunner: 'yarn',
+      installCommand: 'add',
+      lockFile: 'yarn.lock',
+      initDefault: ['init', '--yes'],
+    },
+    pnpm: {
+      name: 'pnpm',
+      executable: 'pnpm',
+      binaryRunner: 'pnpm',
+      installCommand: 'add',
+      lockFile: 'pnpm-lock.yaml',
+      initDefault: ['init'],
+    },
+  };
+
+  /**
+   * Check if a package.json file exists in projectRoot
+   */
+  protected packageJsonExists = (projectRoot: string): boolean => {
+    return this.existsSync(path.resolve(projectRoot, 'package.json'));
+  };
+
+  /**
+   * getPackageManager
+   */
+  getPackageManager() {
+    if (!process.env.npm_config_user_agent) {
+      logger.warn('Could not determine package manager, defaulting to npm');
+      return this.packageManagers['npm'];
+    }
+
+    const userAgent = process.env.npm_config_user_agent;
+    const packageManagerAndVersion = userAgent.split(' ')[0];
+    const packageManagerName = packageManagerAndVersion.split('/')[0];
+
+    if (packageManagerName === 'yarn') {
+      const yarnMajorVersion = packageManagerAndVersion
+        .split('/')[1]
+        .split('.')[0];
+      const yarnName: PackageManagerName = `${packageManagerName}-${
+        yarnMajorVersion === '1' ? 'classic' : 'modern'
+      }`;
+      return this.packageManagers[yarnName];
+    }
+    return this.packageManagers[packageManagerName as PackageManagerName];
+  }
+
+  /**
+   * getPackageManagerController
+   */
+  getPackageManagerController(projectRoot: string) {
```

Comment:
similar comment about `projectRoot` - this one should be taken in constructor and become property in this class. It's invariant during lifecycle of this process.

Reply:
done f95f9b90f